README.md
# Projects
General projects that I am working on currently or have finished


Assignment_01/history.txt
Nicholas Kidd   
    1  cd downloads
    2  cd Downloads
    3  ls
    4  gcc -o prog prog01.c
    5  ./prog 3
    6  ./prog
    7  ./prog 1 2 3
    8  /prog abc
    9  ./prog abc
   10  ./prog abc 1
   11  /prog 1 abc
   12  ./prog 1 abc
   13  ./prog 3 27
   14  ./prog 8 27
   15  ./prog 8 24
   16  ./prog 100 115000
   17  cd Downloads
   18  bash scipt01.sh prog01.c 1
   19  bash script01.sh prog01.c 1
   20  clear
   21  bash script01.sh prog01.c 
   22  bash ./script01.sh ./prog01.c 
   23  ./prog 
   24  bash ./script01.sh ./prog01.c 
   25  man
   26  man cd
   27  man
   28  cd
   29  man -k
   30  ls
   31  cd Desktop
   32  mkdir newFolder
   33  ls
   34  cd newFolder
   35  pwd
   36  clear
   37  man
   38  cd
   39  clear
   40  cd Desktop
   41  mkdir newFolder
   42  cd newFolder
   43  pwd
   44  clear
   45  cd
   46  clear
   47  cd Desktop
   48  mkdir newFolder
   49  cd newFolder
   50  pwd
   51  pwd > path.txt
   52  pr path.txt
   53  touch new.txt
   54  ls
   55  ls -la
   56  rm new.txt
   57  ls
   58  cd Downloads
   59  ls
   60  cd
   61  cd Desktop
   62  ls
   63  cd Screenshots
   64  ls
   65  cd
   66  cd Pictures
   67  ls
   68  history
   69  ls -la
   70  history > history.txt


Prog_02/Scripts/script1.sh
#!/bin/bash

#puts all arguments into a list
arguments=($@)
#doesnt include bash or script1.sh
#determines the amount of arguments
length=0;
for n in ${arguments[@]}; do
    length=$((length+1));
done

#if length is less than 1, no file path present, which exits after displaying error.
if((length < 1))
then
    echo "There is no argument for a file path."
    echo "Proper Usage: script1.sh <pathToFile>."
    exit 1
fi

#if length is more than 1, too many file paths (shoudl only take one), which exits after displaying error.
if((length > 1))
then
    echo "There are too many arguments."
    echo "Proper Usage: script1.sh <pathToFile>."
fi

#if length is == 1 it will do ls -p of the file path which adds a / to the end of all directories and grep -v / does not output anything 
#with a / at the end
ls -p ${arguments[length-1]} | grep -v /


Prog_02/Scripts/script3.sh
#!/bin/bash

gcc -o program ${1}
#puts all arguments into a list
arguments=($@)
#doesnt include bash or script1.sh#
#determines the amount of arguments
length=0;
for n in ${arguments[@]}; do
    length=$((length+1));
done

#if length is less than 1, no file path present, which exits after displaying error.
if((length < 2))
then
    echo "There is no argument for a file path."
    echo "Proper Usage: script1.sh <pathToProgram> <pathToFile>."
    exit 1
fi

 #if length is more than 1, too many file paths (shoudl only take one), which exits after displaying error.
if((length > 2))
then
    echo "There are too many arguments."
    echo "Proper Usage: script1.sh <pathToProgram> <pathToFile>."
fi

#put all dat files found into a variable called dat
allDat=`ls ${2}/*.dat`
#making image list to put all the values into
imageValues=""
#using dat go through all the files and put them into imagelist so that all the coordinates and locaiton names are listed in imagelist
for file in $allDat; do
    contents=`cat $file`
    imageValues="$imageValues $contents"
done

#echo $imagelist
#runs them using prog
./program $imagelist


Assignment_01/Code/prog01.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void noArgument(int argc, char programName[]);
void tooManyArguments(int argc, char programName[]);
void divisorsOneInt(int number);
void calculateGCD(int firstNumber, int secondNumber);

int main(int argc, char* argv[])
{
    int slash=0;
    //Extra Credit
    //Identifies where the slash is in the program name (if there is one) and puts it in slash
    for(int i=(strlen(argv[0])-1);(i>=0)||(slash==0);i--){
        if(argv[0][i]=='/'){
            slash=i;
        }
    }
    //Create new array the length of everything after the slash+1 so there is room for '\0'
    char progName[(strlen(argv[0]))-(slash-1)];
    int slashPtr=slash;
    //starting at the position after the slash, copy the name of the file into progName for    
    //use with errors
    for(int j=0; j<strlen(progName); j++){
        progName[j]=argv[0][slashPtr+1];
        slashPtr++;
    }
    //Manually put null character into it just in case it is not allocated.
    progName[strlen(progName)]='\0';
    //Checks for errors with amount of arguments
    noArgument(argc, progName);
    tooManyArguments(argc, progName);
    //If only one argument besides the executable is entered
    if(argc==2){
        long long int firstValue;
        //Checks to make sure that the argument given is an integer with the sscanf function and that it is
        //strictly postive, if it is not it prints an error and exits the program
        if(sscanf(argv[1], "%lld", &firstValue) !=1 || firstValue<1){
            printf("%s's argument is not a strictly positive integer.\n", progName);
            exit(0);
        }
        //Calls the function to get the divisors for the value
        divisorsOneInt(firstValue);
    //If two arguments besides the executable are entered
    }else if(argc==3){
        long long int firstValue;
        long long int secondValue;
        //Scans to make sure first argument is an interger and strictly positive
        if(sscanf(argv[1], "%lld", &firstValue) !=1 || firstValue <1){
            printf("%s's first argument is not a strictly positive integer.\n", progName);
            exit(0);
        }
        //Scans to make sure second argument is an integer and strictly positive
        if(sscanf(argv[2], "%lld", &secondValue) !=1 || secondValue<1){
            printf("%s's second argument is not a strictly positive integer.\n", progName);
            exit(0);
        }
        //Calls the function to get the GCD of the two numbers
        calculateGCD(firstValue, secondValue);
    }
    return 0;
}

//Takes in the amount of arguments given to the program from argc and the name of the program from argv[0]
//Will use the number of arguments to see if there are enough declared by the user
void noArgument(int argc, char programName[])
{
    //If there are less than two arguments, they either did not run the program or they did not enter a number
    //If this is triggered, it will give an error statement and exit out of the program, but if there are
    //enough arguments, the program will continue as it should
    if(argc < 2){
        printf("%s launched with no argument\n", programName);
        printf("Proper usage: %s m [n]\n", programName);
        exit(0);
    }
}

//Takes in the amount of arguments given to the program from argc and the name of the program from argv[0]
//Uses the number of arguments to see if there were too many declared by the user
void tooManyArguments(int argc, char programName[])
{
    //If there are more than three arguments,they have too many arguments because the program should take
    //the program name and one int to ask for the divisors or the program name and two ints to ask for the
    //the GCD. If this is triggered, it will give an error statement and exit out of the program
    if(argc > 3){
        printf("%s launched with too many arguments\n", programName);
        printf("Proper usage: %s m [n]\n", programName);
        exit(0);
    }
}

//Takes in the number given to the the program and calculate the divisors for the number and prints them out
void divisorsOneInt(int number){
    printf("The list of divisors of %d is: ", number);
    //The program goes through and does the the number %(mod) i which will go from 1 to the number. and if
    //number%1 is equal to 0 (meaning i can equally divide the number), it will print out i.
    for(int i=1;i<=number;i++){
        //will be triggered at the last value of i (which is equal to number) and print i without a space to
        //make the output have correct spacing before the period and line break
        if(number==i){
          printf("%d", i);
        }
        else if(number%i==0){
            printf("%d, ",i);
        }
    }
    //prints out the period and line break after the last divisor.
    printf(".\n");
}

//Takes in the two numbers given to the program and calculates the GCD between them
void calculateGCD(int firstNumber, int secondNumber){
    //All strictly positive integers have a common divisor of 1 so GCD is defaultly set to 1
    int gcd = 1;
    //This will do firstNumber%i and secondNumber%i at the same time and if they are both 0, it will make
    //i the new value of gcd. It will continue to do this until i is bigger than firstNumber or secondNumber
    for(int i=2; i<=firstNumber && i<=secondNumber;i++){
        if(firstNumber%i==0 && secondNumber%i==0){
            gcd=i;
        }
    }
    printf("The gcd of %d and %d is %d.\n", firstNumber, secondNumber, gcd);
}


Prog_02/Programs/prog02_v1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Image {
    int x; // x-coordinate
    int y; //y-coordinate
    char locationName[100]; //name of the location
    int imageNumber; //number of entry (image xxx)
} Image;

// void isCharacter(char loc[], int index);
void checkQuery(char query[], char subject[], Image DataList[], int index);
void searchX(Image DataList[], char subject[], int index);
void searchY(Image DataList[], char subject[], int index);
void searchN(Image DataList[], char subject[], int index);
// void isCharacterQuery(char loc[]);
void printData(Image DataList[], int i);

int main(int argc, char* argv[]){
  //index of how many structs in DataList array
    int index = 0;
    //If there are less than 4 arguments, there are not enough arguments to compute anything.
    if(argc < 4){
        printf("Invalid arguments. try ./<program name> <x coordinate> <y coordinate> <location>.\n");
        exit(0);
    }
    //initializes array of Images
    Image DataList[argc];
    //starting at 1 (0 is ./prog), increment i 3 times because each image is i, i+1 and i+2
    for(int i = 1; i < argc;i += 3){
      //temps for error checking and assigning into structs
        int xValue;
        int yValue;
        char locationValue[1000];
        //if x coordinate is not an integer, ERROR
        if(sscanf(argv[i], "%d", &xValue) != 1){
            printf("The x coordinate in entry number %d is not a valid integer.\n", index);
            exit(0);
        }
        //if y coordinate is not an integer, ERROR
        else if(sscanf(argv[i+1], "%d", &yValue) != 1){
            printf("The y coordinate in entry number %d is not a valid integer.\n", index);
            exit(0);
        }
        //if location is NULL, ERROR
        if(argv[i+2] == NULL){
          printf("The location in entry number %d is not a valid location name.\n", index);
          exit(0);
        }
        //test if location is a string using sscanf to put it into variable and my function isCharacter to check
        sscanf(argv[i+2],"%s", locationValue);
        // isCharacter(locationValue, index);
        //assign each element in the struct to the temp values.
        strcpy(DataList[index].locationName, locationValue);
        DataList[index].x = xValue;
        DataList[index].y = yValue;
        DataList[index].imageNumber = index;
        index++;
    }

    // used to be my function for printing out all the data while it is all in the DataList struct
    // for(int i = 0; i<index;i++){
    //     printData(DataList, i);
    // }
    //function given to us by Herve with my own touch
    int keepGoing = 1;
    while (keepGoing)
    {
        char line[200];
	//	the [^\n] means "any character other than end of line.""  If I wanted, say, other
	//	than tab or end of line, it would be [^\t\n]
	      scanf("%[^\n]", line);
	//	The \n character is still on the standard input, so I need to "consume" it, otherwise
	//	I would keep reading empty lines.
	      getc(stdin);
	//printf("line: -->|%s|<--\n", line);
        //printf("length: %ld", strlen(line));
	      if (strcmp(line, "EXIT") == 0)
	          keepGoing = 0;
	      else
	      {
           // subject is search query, query is the lab -x, -y, -n, etc.
	          char subject[200];
            char query[3];
            //checks the query and the subject
            sscanf(line,"%s %s", query, subject);
            //makes sure that there is not more than a query and a subject, if there is, its an error
            if(line[(strlen(query))+(strlen(subject))+1] != '\0'){
                printf("The query is not valid.\n");
                exit(0);
            }
            //check if query is right.
            checkQuery(query, subject, DataList, index);
	}
    }
    return 0;
}


//function to test if the string loc[] is a valid char array of letters and/or _
//TOLD BY HERVE TO NOT INCLUDE BECAUSE THE NAME CAN CONTAIN ANYTHING BUT A SPACE
// void isCharacter(char loc[], int index){
//   //goes char by char for length of loc and compares it to a through z, A through Z, and _
//     for (int i = 0; i < strlen(loc); i++){
//         if(('a' <= loc[i] && loc[i] <= 'z') || ('A' <= loc[i] && loc[i] <= 'Z') || (loc[i] == '_')){
//         } else{
//           //if not valid, it will output this error
//             printf("The location name in entry number %d is not a valid location name\n", index);
//             exit(0);
//         }
//     }
// }

//function to test if it is a character specifically for the queries (different error message)
//TOLD BY HERVE TO NOT INCLUDE BECAUSE THE NAME CAN CONTAIN ANYTHING BUT A SPACE
// void isCharacterQuery(char loc[]){
//     for (int i = 0; i < strlen(loc); i++){
//         if(('a' <= loc[i] && loc[i] <= 'z') || ('A' <= loc[i] && loc[i] <= 'Z') || (loc[i] == '_')){
//         } else{
//             printf("The query is not valid, please enter a location name after -n.\n");
//             exit(0);
//         }
//     }
// }

//a function to check the query given by the user. detects if query is valid and if it is forwards the subject with
//necessary info to another helper function that computes the query
void checkQuery(char query[], char subject[], Image DataList[], int index){
    //queries -x -y -n all start with -
    if(query[0] == '-'){
    } else{
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
    //all queries -x -y -n end with \0 on second index
    if(query[2] == '\0'){
    } else{
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
    //detects if query[1] is x y or z and sends to helped functions
    if(query[1] == 'x'){
        searchX(DataList, subject, index);
    }else if(query[1] == 'n'){
        searchN(DataList, subject, index);
    }else if(query[1] == 'y'){
        searchY(DataList, subject, index);
    } else {
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
}

//prints the data and computes the accurate image number under 1000 <- limitation that can be changed based on format
//showed in pdf of image number.
void printData(Image DataList[], int i){
    if(DataList[i].imageNumber < 10){
        printf("image 00%d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber %10,DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber < 100) {
        printf("image 0%d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber %100, DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber < 1000) {
        printf("image %d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber %1000, DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber >= 1000) {
        printf("image %d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber, DataList[i].x, DataList[i].y, DataList[i].locationName);
    }
}

//searches for a location based on the name of the city
void searchN(Image DataList[], char subject[], int index){
    // checks if subject is a character
    // isCharacterQuery(subject);
    //documents noneMatch so if there is no matching location it outputs that there is no matching image data
    int noneMatch = 0;
    for(int i = 0; i < index; i++){
      // same to make sure that all the chars are equal
        int same = 1;
        if ((strlen(subject)) == (strlen(DataList[i].locationName))){
            for(int j = 0; (j < strlen(subject)) && (same == 1); j++){
                if(DataList[i].locationName[j] == subject[j]){
                    same = 1;
                }else {
                    same = 0;
                }
            }//if same == 1 locationName is the same a subjecy and print data for the locatonName by going to printData
            if(same==1){
                noneMatch=1;
                printData(DataList, i);
            }
        }
    }
    if(noneMatch == 0){
        printf("No image found at that location.\n");
    }
}

//function that searches DataList by x coordinates
void searchX(Image DataList[], char subject[], int index){
     //temp var for the subject and makes sure it is an int
    int coordinate;
    if(sscanf(subject, "%d", &coordinate) !=  1){
        printf("The x coordinate to search for is not a valid integer.\n");
        exit(0);
    }
    //used the same as noneMatch in searchN
    int noneMatch = 0;
    ///if coordinate==x in datalist output data
    for(int i = 0; i < index; i++){
        if(coordinate == DataList[i].x){
            printData(DataList, i);
            noneMatch = 1;
        }
    }
    if(noneMatch == 0){
        printf("No image found at that location.\n");
    }
}

//function that searches data list by y coordinates
void searchY(Image DataList[], char subject[], int index){
//used the same as in searchX
    int coordinate;
    if(sscanf(subject, "%d", &coordinate) !=  1){
        printf("The y coordinate to search for is not a valid integer.\n");
        exit(0);
    }
    //used the as in searchX
    int noneMatch = 0;
    //if coordinate == data list y coordinate, output data
    for(int i = 0; i < index; i++){
        if(coordinate == DataList[i].y){
            printData(DataList, i);
            noneMatch = 1;
        }
    }
    if(noneMatch == 0){
        printf("No image found at that location.\n");
    }
}


Assignment_01/Code/script01.sh
#!/bin/bash

#compiles the code
gcc -o prog ${1}

#puts all arguments into a list
arguments=($@)

#determines the amount of arguments
length=0;
for n in ${arguments[@]:1}; do
    length=$((length+1));
done

#if length is less than one, there are not enough arguments 
if((length<1))
then
    echo "Invalid argument list. Not enough arguments are present."
    echo "Proper usage: script01 <path to source file> m1 [m2 [m3 [...]]]"
    exit 1
fi

#for each argument test to see if it is strictly a positive integer, if not, display error and exit
for i in ${arguments[@]:1};
do
    if (($i < 1))
    then
        echo "Invalid arguent list. All arguments should be strictly positive integers."
        echo "Proper usage: script01 <path to source file> m1 [m2 [m3 [...]]]"
        exit 1
    fi
done

#after all error checking run the program to get the divisors of each argument (except for name of program)
for i in ${arguments[@]:1};
do
    ./prog "$i"
done


#run the program to get the gcd of all of the arguments in the order they are presented
for ((j=1; j<=length; j++)); do
     for ((k=j+1;k<=length;k++)); do
             ./prog ${arguments[j]} ${arguments[k]}
     done
done 




Prog_02/Programs/prog02_v2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Image {
    int x; // x-coordinate
    int y; //y-coordinate
    char locationName[100]; //name of the location
    int imageNumber; //number of entry (image xxx)
} Image;

typedef struct loc{
    char city[100]; //city name
    int indices[1000]; //list of indexes containing Data with city name
    int idx; //Next available index of indices list
} loc;

// void isCharacter(char loc[], int index);
void checkQuery(char query[], char subject[], Image DataList[], int index, loc LocationList[], int locIndex);
void searchX(Image DataList[], char subject[], int index);
void searchY(Image DataList[], char subject[], int index);
void searchN(loc LocationList[], char subject[], int locIndex, Image DataList[]);
// void isCharacterQuery(char loc[]);
void printData(Image DataList[], int i);
void printDataLoc(loc LocationList[], int i, Image DataList[]);

int main(int argc, char* argv[]){
    //index of how many structs in DataList array
    int index = 0;
    //index of how many structs in LocationList array
    int locIndex = 0;
    //If there are less than 4 arguments, there are not enough arguments to compute anything.
    if(argc < 4){
        printf("Invalid arguments. try ./<program name> <x coordinate> <y coordinate> <location>.\n");
        exit(0);
    }
    //initializes array of Images and array of locs(locations)
    Image DataList[argc];
    //had to hardcode 1500 because it was giving me a seg fault if both LocationList and DataList had argc elements
    loc LocationList[1500];
    //starting at 1 (0 is ./prog), increment i 3 times because each image is i, i+1 and i+2
    for(int i = 1; i < argc;i += 3){
        //temps for error checking and assigning into structs
        int xValue;
        int yValue;
        char locationValue[100];
        //if x coordinate is not an integer, ERROR
        if(sscanf(argv[i], "%d", &xValue) != 1){
            printf("The x coordinate in entry number %d is not a valid integer.\n", index);
            exit(0);
        }
        //if y coordinate is not an integer, ERROR
        else if(sscanf(argv[i+1], "%d", &yValue) != 1){
            printf("The y coordinate in entry number %d is not a valid integer.\n", index);
            exit(0);
        }
        //if location is NULL, ERROR
        if(argv[i+2] == NULL){
          printf("The location in entry number %d is not a valid location name.\n", index);
          exit(0);
        }
        //test if location is a string using sscanf to put it into variable, used to use my function isCharacter to check
        sscanf(argv[i+2],"%s", locationValue);
        // isCharacter(locationValue, index);
        //assign each element in the struct to the temp values.
        strcpy(DataList[index].locationName, locationValue);
        DataList[index].x = xValue;
        DataList[index].y = yValue;
        DataList[index].imageNumber = index;
        //if there are no images, there are no locs so initialize the first loc
        if(index==0){
           //copy name of city from locationName
            strcpy(LocationList[0].city, DataList[0].locationName);
            //The first element in the list of indices is 0
            LocationList[0].indices[0] = index;
            //Next index is 1
            LocationList[0].idx = 1;
            locIndex++;
        }else{
          // a variable to test if the locationValue is already in the list of loc structs
            int found = 0;
            for(int j = 0; (j < locIndex) && (found == 0); j++){
                //check if the length of the city and locationValue are the same, if so check if the strings are same using
                //strcmp
                if ((strlen(LocationList[j].city)) == (strlen(locationValue))){
                    if(strcmp(LocationList[j].city,locationValue)==0){
                        LocationList[j].indices[LocationList[j].idx] = index;
                        LocationList[j].idx++;
                        found=1;
                    }
                }
            }
            //if there is no struct found with the same city name, it will initalize a new struct that with the data.
            if(found==0){
                strcpy(LocationList[locIndex].city, DataList[index].locationName);
                LocationList[locIndex].indices[0] = index;
                LocationList[locIndex].idx = 1;
                locIndex++;
            }
        }
        index++;
    }
    // used to be my function for printing out all the data while it is all in the DataList struct
    // for(int i = 0; i<index;i++){
    //     printData(DataList, i);
    // }
    //reading user's input given to us by Herve
    int keepGoing = 1;
    while (keepGoing)
    {
        char line[200];
	//	the [^\n] means "any character other than end of line.""  If I wanted, say, other
	//	than tab or end of line, it would be [^\t\n]
	      scanf("%[^\n]", line);
	//	The \n character is still on the standard input, so I need to "consume" it, otherwise
	//	I would keep reading empty lines.
	      getc(stdin);
	//printf("line: -->|%s|<--\n", line);
        //printf("length: %ld", strlen(line));
	   if (strcmp(line, "EXIT") == 0)
	       keepGoing = 0;
	   else
	   {
           // subject is search query, query is the lab -x, -y, -n, etc.
	          char subject[200];
            char query[5];
            //checks the query and the subject
            sscanf(line,"%s %s", query, subject);
            //makes sure that there is not more than a query and a subject, if there is, its an error
            if(line[(strlen(query))+(strlen(subject))+1] != '\0'){
                printf("The query is not valid.\n");
                exit(0);
            }
            //check if query is right.
            checkQuery(query, subject, DataList, index, LocationList, locIndex);
	}
    }
    return 0;
}

//function to test if the string loc[] is a valid char array of letters and/or _
//TOLD BY HERVE TO NOT INCLUDE BECAUSE THE NAME CAN CONTAIN ANYTHING BUT A SPACE
// void isCharacter(char loc[], int index){
//     //goes char by char for length of loc and compares it to a through z, A through Z, and _
//     for (int i = 0; i < strlen(loc); i++){
//         if(('a' <= loc[i] && loc[i] <= 'z') || ('A' <= loc[i] && loc[i] <= 'Z') || (loc[i] == '_')){
//         } else{
//             //if not valid, it will output this error
//             printf("The location name in entry number %d is not a valid location name\n", index);
//             exit(0);
//         }
//     }
// }

//function to test if it is a character specifically for the queries (different error message)
//TOLD BY HERVE TO NOT INCLUDE BECAUSE THE NAME CAN CONTAIN ANYTHING BUT A SPACE
// void isCharacterQuery(char loc[]){
//     for (int i = 0; i < strlen(loc); i++){
//         if(('a' <= loc[i] && loc[i] <= 'z') || ('A' <= loc[i] && loc[i] <= 'Z') || (loc[i] == '_')){
//         } else{
//             printf("The query is not valid, please enter a location name after -n.\n");
//             exit(0);
//         }
//     }
// }

//a function to check the query given by the user. detects if query is valid and if it is forwards the subject with
//necessary info to another helper function that computes the query
void checkQuery(char query[], char subject[], Image DataList[], int index, loc LocationList[], int locIndex){
  //queries -x -y -n all start with -
    if(query[0] == '-'){
    } else{
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
    //all queries -x -y -n end with \0 on second index
    if(query[2] == '\0'){
    } else{
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
    //detects if query[1] is x y or z and sends to helped functions
    if(query[1] == 'x'){
        searchX(DataList, subject, index);
    }else if(query[1] == 'n'){
        searchN(LocationList, subject, locIndex, DataList);
    }else if(query[1] == 'y'){
        searchY(DataList, subject, index);
    } else {
        printf("The query is not valid, try -x <number>, -y <number>, or -n <name>.\n");
        exit(0);
    }
}

//prints the data and computes the accurate image number under 1000 <- limitation that can be changed based on format
//showed in pdf of image number.
void printData(Image DataList[], int i){
    if(DataList[i].imageNumber < 10){
        printf("image 00%d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber,DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber < 100) {
        printf("image 0%d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber, DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber < 1000) {
        printf("image %d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber, DataList[i].x, DataList[i].y, DataList[i].locationName);
    } else if(DataList[i].imageNumber >= 1000) {
        printf("image %d. location: %d, %d. Name: %s.\n", DataList[i].imageNumber, DataList[i].x, DataList[i].y, DataList[i].locationName);
    }
}

//takes in loc information and goes through the indices array in the loc to direct what image struct to print data from
void printDataLoc(loc LocationList[], int i, Image DataList[]){
    for(int j = 0; j < LocationList[i].idx; j++){
        printData(DataList, LocationList[i].indices[j]);
    }
}

//searches for a location based on the name of the city
void searchN(loc LocationList[], char subject[], int locIndex, Image DataList[]){
    // checks if subject is a character
    // isCharacterQuery(subject);
    //documents noneMatch so if there is no matching location it outputs that there is no matching image data
    int noneMatch = 0;
    for(int i = 0; i < locIndex; i++){
        // same to make sure that all the chars are equal
        int same = 1;
        if ((strlen(subject)) == (strlen(LocationList[i].city))){
            for(int j = 0; (j < strlen(subject)) && (same == 1); j++){
                if(LocationList[i].city[j] == subject[j]){
                    same = 1;
                }else {
                    same = 0;
                }
            }
            //if same == 1 city is the same a subjecy and print data for the city by going to printDataLoc which directs
            //to printData
            if(same==1){
                noneMatch=1;
                printDataLoc(LocationList, i, DataList);
            }
        }
    }
    if(noneMatch == 0){
        printf("No images found at that location.\n");
    }
}

//function that searches DataList by x coordinates
void searchX(Image DataList[], char subject[], int index){
    //temp var for the subject and makes sure it is an int
    int coordinate;
    if(sscanf(subject, "%d", &coordinate) !=  1){
        printf("The x coordinate to search for is not a valid integer.\n");
        exit(0);
    }
    //used the same as noneMatch in searchN
    int noneMatch = 0;
    ///if coordinate==x in datalist output data
    for(int i = 0; i < index; i++){
        if(coordinate == DataList[i].x){
            printData(DataList, i);
            noneMatch = 1;
        }
    }
    if(noneMatch == 0){
        printf("No images found at that location.\n");
    }
}

//function that searches data list by y coordinates
void searchY(Image DataList[], char subject[], int index){
    //used the same as in searchX
    int coordinate;
    if(sscanf(subject, "%d", &coordinate) !=  1){
        printf("The y coordinate to search for is not a valid integer.\n");
        exit(0);
    }
    //used the as in searchX
    int noneMatch = 0;
    //if coordinate == data list y coordinate, output data
    for(int i = 0; i < index; i++){
        if(coordinate == DataList[i].y){
            printData(DataList, i);
            noneMatch = 1;
        }
    }
    if(noneMatch == 0){
        printf("No images found at that location.\n");
    }
}
